File: Java/queue.java
package Java;
import java.util.Queue;
import java.util.LinkedList;

public class queue {
    public static void main(String[] args) {
        
        // Queue = FIFO data structure. First-In First_out (ex. A line of people)
        //         A collection designed for holding elements prior to processing
        //         Linear data structure

        //         add    = enqueue, offer()
        //         remove = dequeue, poll()

        Queue<String> queue = new LinkedList<String>();

        // System.out.println(queue.isEmpty());
        
        queue.offer("Karen");
        queue.offer("Chad");
        queue.offer("Steve");
        queue.offer("Harold");

        // System.out.println(queue.isEmpty());
        // System.out.println(queue.size());
        System.out.println(queue.contains("Harold"));

        // System.out.println(queue.peek());
        // queue.poll();
        // queue.poll();
        

        System.out.println(queue);

        // Where are queues useful?
        // 1. Keyboard Buffer (letters should appear on the screen in the orderr they're pressed)
        // 2. Printer Queue (Print jobs should be completed in order)
        // 3. Used in LinkedLists, PriorityQueues, Breadth-first search
    }
}

File: Python/queue.py
class ListQueue:
    def __init__(self):
        self._items = []
    def enQueue(self, item):
        self._items.append(item)
    def deQueue(self):
        return self._items.pop(0)
    def peek(self):
        return self._items[0]

class PointerQueue:
    def __init__(self):
        self._items = []
        self._front = 0

    def deQueue(self):
        item = self._items[self._front]
        self._front += 1
        return item
    
queue = ListQueue()

queue.enQueue(1)
print(queue.peek())

queue.enQueue(3)
print(queue.peek())

queue.deQueue()
print(queue.peek())


class PointerQueue:
    def __int__(self):
        self._items = []
        self._front = 0

    def deQueue(self):
        item = self._items[self._front]
        self._front += 1
        return item